libyul/YulName.cpp

```diff
@@ -0,0 +1,435 @@
+
+			if (auto const* equalityFunction = m_dialect.get().equalityFunction(YulString(typeLabel)))
+				m_predefinedBuiltinFunctions.equalityFunctions.emplace_back(nameOfBuiltin(equalityFunction->name.str()));
+			else
+				m_predefinedBuiltinFunctions.equalityFunctions.emplace_back(std::nullopt);
+
+			if (auto const* booleanNegationFunction = m_dialect.get().booleanNegationFunction())
+				m_predefinedBuiltinFunctions.booleanNegationFunction = nameOfBuiltin(booleanNegationFunction->name.str());
+			else
+				m_predefinedBuiltinFunctions.booleanNegationFunction = std::nullopt;
+
+			if (auto const* memStoreFunction = m_dialect.get().memoryStoreFunction(YulString(typeLabel)))
+				m_predefinedBuiltinFunctions.memoryStoreFunctions.emplace_back(nameOfBuiltin(memStoreFunction->name.str()));
+			else
+				m_predefinedBuiltinFunctions.memoryStoreFunctions.emplace_back(std::nullopt);
+
+			if (auto const* memLoadFunction = m_dialect.get().memoryLoadFunction(YulString(typeLabel)))
+				m_predefinedBuiltinFunctions.memoryLoadFunctions.emplace_back(nameOfBuiltin(memLoadFunction->name.str()));
+			else
+				m_predefinedBuiltinFunctions.memoryLoadFunctions.emplace_back(std::nullopt);
+
+			if (auto const* storageStoreFunction = m_dialect.get().storageStoreFunction(YulString(typeLabel)))
+				m_predefinedBuiltinFunctions.storageStoreFunctions.emplace_back(nameOfBuiltin(storageStoreFunction->name.str()));
+			else
+				m_predefinedBuiltinFunctions.storageStoreFunctions.emplace_back(std::nullopt);
+
+			if (auto const* storageLoadFunction = m_dialect.get().storageLoadFunction(YulString(typeLabel)))
+				m_predefinedBuiltinFunctions.storageLoadFunctions.emplace_back(nameOfBuiltin(storageLoadFunction->name.str()));
+			else
+				m_predefinedBuiltinFunctions.storageLoadFunctions.emplace_back(std::nullopt);
+
+			m_predefinedBuiltinFunctions.hashFunctions.emplace_back(nameOfBuiltin(m_dialect.get().hashFunction(YulString(typeLabel)).str()));
+		}
+	}
+
+	m_predefined.placeholder_zero = defineName("@ 0");
+	m_predefined.placeholder_one = defineName("@ 1");
+	m_predefined.placeholder_thirtytwo = defineName("@ 32");
+}
+
+YulNameRepository::BuiltinFunction const* YulNameRepository::builtin(YulName const _name) const
+{
+	yulAssert(nameWithinBounds(_name), "YulName exceeds repository size, probably stems from another instance.");
+	auto const baseName = baseNameOf(_name);
+	if (isBuiltinName(baseName))
+	{
+		auto const it = m_builtinFunctions.find(_name);
+		if (it != m_builtinFunctions.end())
+			return &it->second;
+		else
+			return nullptr;
+	}
+	return nullptr;
+}
+
+std::optional<std::string_view> YulNameRepository::labelOf(YulName const _name) const
+{
+	yulAssert(nameWithinBounds(_name), "YulName exceeds repository size, probably stems from another instance.");
+	if (!isDerivedName(_name))
+	{
+		// if the parent is directly a defined label, we take that one
+		return m_definedLabels[std::get<0>(m_names[_name])];
```

Comment:
Is it guarantee that `std::get<0>(m_names[_name])` will never return an invalid index to the `m_definedLabels` vector? I'm assuming that the size of `m_names` and `m_definedLabels` will always be the same, so maybe we could assert that.

Reply:
It's guaranteed by how the datastructure works, so asserting it's within bounds seems like the right thing to do:)